src/aura/BGE_DataImportBatchEntry/BGE_DataImportBatchEntryHelper.js

```diff
@@ -213,6 +267,60 @@
         $A.enqueueAction(action);
     },
 
+    /**
+     * @description: insert or update record and possible error into the table data
+     * @param updatedRecord: DataImport__c record that needs to be updated in the table
+     */
+    afterDryRun: function (component, updatedRecord, isNewRecord) {
+        let tableRow = this.flattenDataImportRow(updatedRecord);
+        let hasError = false;
+
+        //get payment and opportunity error information if import failed
+        if (tableRow.errors.length > 0) {
+            hasError = true;
+            let rowError = {
+                id: tableRow.Id,
+                title: $A.get('$Label.c.PageMessagesError'),
+                messages: tableRow.errors
+            };
+            this.handleNewTableErrors(component, [rowError]);
+        }
+
+        if (isNewRecord) {
+            this.insertTableRow(component, tableRow);
+        } else {
+            this.replaceTableRow(component, tableRow);
+            if (!hasError) {
+                this.clearOldTableErrors(component, tableRow.Id);
+            }
+        }
+    },
+
+    /**
+     * @description: finds the index of the row and replaces it with the updated row
+     * @param tableRow: flattened row for insert into the table
+     */
+    replaceTableRow: function(component, tableRow) {
+        let rows = component.get('v.data');
+        for (let i = 0; i < rows.length; i++) {
+            if (rows[i].Id === tableRow.Id) {
```

Comment:
Yes, I really wanted a better way to do this too, and gave it some thought. The issue is that when a user makes a change, onCellChange is fired and only gives us the draftValues. This is the Id of the row and the field that has changed by API name. 

In contrast, when I delete something, I can match for the indexOf in the table, because handleRowAction passes me the entire row and I can get an exact match on it.

Maybe there's a way to have something running in JS at the same time the action is firing? I'm a little nervous about race conditions, but that would make things more efficient.

Reply:
Did a little further testing in onCellChange and confirmed that's all we have to work with:

`console.log(JSON.parse(JSON.stringify(event.getParams())));
{draftValues: Array(1)}
   draftValues: Array(1)
      0: {npsp__Donation_Amount__c: "16", Id: "a0S5C000006ugREUAY"}
      length: 1
   __proto__: Array(0)
__proto__: Object
`